core/src/main/scala/caliban/Value.scala

```diff
@@ -201,6 +210,176 @@ private object ValueCirce {
+        Value.FloatValue.BigDecimalNumber(in.readBigDecimal(null))
+      def encodeValue(x: Value.FloatValue.BigDecimalNumber, out: JsonWriter): Unit =
+        out.writeVal(x.value)
+      def nullValue: Value.FloatValue.BigDecimalNumber =
+        null
+    }
+
+  implicit val stringValueCodec: JsonValueCodec[Value.StringValue] =
+    new JsonValueCodec[Value.StringValue] {
+      def decodeValue(in: JsonReader, default: Value.StringValue): Value.StringValue =
+        Value.StringValue(in.readString(null))
+      def encodeValue(x: Value.StringValue, out: JsonWriter): Unit =
+        out.writeVal(x.value)
+      def nullValue: Value.StringValue =
+        null
+    }
+
+  implicit val enumValueCodec: JsonValueCodec[Value.EnumValue] =
+    new JsonValueCodec[Value.EnumValue] {
+      def decodeValue(in: JsonReader, default: Value.EnumValue): Value.EnumValue =
+        Value.EnumValue(in.readString(null))
+      def encodeValue(x: Value.EnumValue, out: JsonWriter): Unit =
+        out.writeVal(x.value)
+      def nullValue: Value.EnumValue =
+        null
+    }
+
+  implicit val listValueCodec: JsonValueCodec[ResponseValue.ListValue] =
+    new JsonValueCodec[ResponseValue.ListValue] {
+      def decodeValue(in: JsonReader, default: ResponseValue.ListValue): ResponseValue.ListValue =
+        if (in.isNextToken('[')) { // TODO test this
+          val values = List.newBuilder[ResponseValue]
+          if (!in.isNextToken(']')) {
+            in.rollbackToken()
+            var l = -1
+            while (l < 0 || in.isNextToken(',')) {
+              l = 0
+              val responseValue = responseValueCodec.decodeValue(in, responseValueCodec.nullValue)
+              values += responseValue
+            }
+          }
+          ResponseValue.ListValue(values.result)
+        } else {
+          in.readNullOrTokenError(default, '[')
+        }
+      def encodeValue(x: ResponseValue.ListValue, out: JsonWriter): Unit = {
+        out.writeArrayStart()
+        x.values.foreach(x1 => responseValueCodec.encodeValue(x1, out))
+        out.writeArrayEnd()
+      }
+      def nullValue: ResponseValue.ListValue =
+        null
+    }
+
+  implicit val objectValueCodec: JsonValueCodec[ResponseValue.ObjectValue] =
+    new JsonValueCodec[ResponseValue.ObjectValue] {
+      def decodeValue(in: JsonReader, default: ResponseValue.ObjectValue): ResponseValue.ObjectValue =
+        if (in.isNextToken('{')) { // TODO test this
+          val values: mutable.Buffer[(String, ResponseValue)] = mutable.ListBuffer[(String, ResponseValue)]()
+          if (!in.isNextToken('}')) {
+            in.rollbackToken()
+            var l = -1
+            while (l < 0 || in.isNextToken(',')) {
+              l = 0
+              val key           = in.readKeyAsString()
+              val responseValue = responseValueCodec.decodeValue(in, responseValueCodec.nullValue)
+              values.append(key -> responseValue)
+            }
```

Comment:
As an option instead of a flag you can use the following loop:
```
            while ({
              values.append((in.readKeyAsString(), responseValueCodec.decodeValue(in, responseValueCodec.nullValue)))
              in.isNextToken(',')
            })
```

Reply:
nifty.